src/include/ipc2/manifest.h

```diff
@@ -0,0 +1,100 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2020 Intel Corporation. All rights reserved.
+ *
+ */
+
+#ifndef SRC_INCLUDE_IPC2_MANIFEST_H_
+#define SRC_INCLUDE_IPC2_MANIFEST_H_
+
+#include <stdint.h>
+#include <kernel/ext_manifest.h>
+
+/*
+ * IPC2 uses the extended manifest to tell the kernel about the supported
+ * firmware features and IPC ABI. This way the kernel knows exactly how to
+ * communicate directly with firmware for each feature and use case for
+ * each version of firmware.
+ *
+ * The extended manifest contains the following IPC2 information that is
+ * used by the drivers
+ *
+ * 1) A list of features classes - this is used by the core driver to
+ *    enumerate feature drivers like audio, sensing, shell, debug, etc.
+ *    i.e. if feature class is listed in the manifest then the core driver
+ *    should enumerate that feature driver. The feature driver can then
+ *    parse the manifest data relevant to itself.
+ *
+ * 2) A list of sub classes per feature - this is used by the feature driver
+ *    to determine the feature sub classes supported by the firmware. e.g for
+ *    audio, the feature driver sub classes could be PCMs, controls, media
+ *    decode, WoV, etc.
+ *
+ * 3) A list of actions per sub class - this is used by the feature driver to
+ *    determine the ABI used for each sub class action. This includes
+ *    mandatory and optional parameters for each sub class action so that
+ *    feature driver can build the correct IPC per ABI version.
+ */
+
+/*
+ * IPC2 Action
+ *
+ * Contains a list of optional and mandatory tuples needed by the driver
+ * to perform this action. Tuples are uint16_t. Structure is aligned on
+ * word.
+ */
+struct sof_ipc2_action {
```

Comment:
are there going to be helpers to help traverse these items? it seems you need to go all the way to the bottom to get the next top level element.

Reply:
Good point - we can currently get to next class, subclass and action by adding "size" to the current ptr.  The use case would be the :
1) common high level code would read each class ID only and send the class data to each driver - e.g. audio class to audio driver, sensing class to sensor driver.
2) feature driver (like audio) would then read each subclass ID and forward each subclass to the correct feature logic to parse the actions i.e. PCM subclass to pcm code, control subclass to kcontrol code.
3) use case file (like pcm.c) parses each action